File: src/main/java/com/nightcat/notice/service/NoticeService.java
package com.nightcat.notice.service;

import com.nightcat.common.CatException;
import com.nightcat.entity.Notice;
import com.nightcat.event.Event;
import com.nightcat.event.EventExecutor;
import com.nightcat.event.EventManager;
import com.nightcat.repository.NoticeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.sql.Timestamp;

import static com.nightcat.common.utility.Util.now;
import static com.nightcat.common.utility.Util.uuid;

@Service
public class NoticeService {

    @Autowired
    private NoticeRepository noticeRepository;

    public NoticeSender sender() {
        return new NoticeSender(noticeRepository);
    }

    @Autowired
    private EventManager eventManager;

    public void init() {
        eventManager.register(Event.ProjectPublishedEvent, projectPublishEventExector());
    }

    public static class NoticeSender {

        private String id = uuid();
        private String content;
        private String uid;
        private Timestamp create_time = now();
        private boolean read = false;
        private Notice.Type type;
        private boolean del = false;

        private NoticeRepository noticeRep;

        NoticeSender(NoticeRepository noticeRep) {
            this.noticeRep = noticeRep;
        }

        public NoticeSender content(String content) {
            this.content = content;
            return this;
        }

        public NoticeSender uid(String uid) {
            this.uid = uid;
            return this;
        }


        public NoticeSender type(Notice.Type type) {
            this.type = type;
            return this;
        }

        public void send() {

            if (uid == null) {
                throw new CatException("NoticeSender uid must be exist");
            }

            if (type == null) {
                throw new CatException("NoticeSender type must be exist");
            }

            if (content == null) {
                throw new CatException("NoticeSender content must be exist");
            }

            Notice notice = new Notice(id, content, uid,
                    create_time, read, type, del);

            this.noticeRep.save(notice);
        }
    }

    public EventExecutor projectPublishEventExector() {
        return new EventExecutor() {
            @Override
            public void execute(Event event, Object context) {
                if (event != Event.ProjectPublishedEvent) return;

            }
        };
    }
}

File: src/main/kotlin/com/yemao/notice/service/notice.service.kt
package com.yemao.notice.service

import com.yemao.common.CatException
import com.yemao.common.base.BaseObject
import com.yemao.notice.models.Notice
import com.yemao.notice.repository.NoticeRepository
import com.yemao.utility.Util.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class NoticeService : BaseObject() {

    @Autowired lateinit var noticeRepository: NoticeRepository

    fun sender(): NoticeSender {
        return NoticeSender(noticeRepository)
    }

    fun findByUid(uid: String): List<Notice> {
        return noticeRepository.findByUid(uid)
    }


    fun findById(id: String): Notice? {
        return noticeRepository.findById(id)
    }

    fun read(notice: Notice) {
        notice.read = true
        noticeRepository.update(notice)
    }

    fun del(notice: Notice) {
        notice.del = true
        noticeRepository.update(notice)
    }


    class NoticeSender internal constructor(private val noticeRep: NoticeRepository) {

        private val id = uuid()
        private var content: String? = null
        private var uid: String? = null
        private val create_time = now()
        private val read = false
        private var type: Notice.Type = Notice.Type.SYSTEM
        private val del = false

        fun content(content: String): NoticeSender {
            this.content = content
            return this
        }

        fun uid(uid: String): NoticeSender {
            this.uid = uid
            return this
        }


        fun type(type: Notice.Type): NoticeSender {
            this.type = type
            return this
        }

        fun send() {

            if (uid == null) {
                throw CatException("NoticeSender uid must be exist")
            }

            if (content == null) {
                throw CatException("NoticeSender content must be exist")
            }

            val userNotice = Notice(id, content as String, uid as String,
                    create_time, read, type, del)

            this.noticeRep.save(userNotice)
        }
    }
}